src/protocolsupportpocketstuff/skin/PeToPcProvider.java

```diff
@@ -1,46 +1,67 @@
 package protocolsupportpocketstuff.skin;
 
 import java.util.Base64;
+import java.util.Map;
+import java.util.WeakHashMap;
 
 import org.bukkit.event.EventHandler;
 import org.bukkit.event.Listener;
 
 import protocolsupport.api.Connection;
+import protocolsupport.api.events.PlayerLoginFinishEvent;
 import protocolsupport.api.events.PlayerProfileCompleteEvent;
+import protocolsupport.api.utils.ProfileProperty;
+import protocolsupport.protocol.utils.authlib.GameProfile;
 import protocolsupportpocketstuff.ProtocolSupportPocketStuff;
 import protocolsupportpocketstuff.api.skins.SkinUtils;
 import protocolsupportpocketstuff.api.skins.SkinUtils.SkinDataWrapper;
 import protocolsupportpocketstuff.api.util.PocketPacketHandler;
 import protocolsupportpocketstuff.api.util.PocketPacketListener;
 import protocolsupportpocketstuff.packet.handshake.ClientLoginPacket;
 import protocolsupportpocketstuff.packet.play.SkinPacket;
+import protocolsupportpocketstuff.storage.Skins;
+import protocolsupportpocketstuff.util.StuffUtils;
 import protocolsupportpocketstuff.util.PacketUtils;
 
+
 public class PeToPcProvider implements PocketPacketListener, Listener {
 
 	private static final ProtocolSupportPocketStuff plugin = ProtocolSupportPocketStuff.getInstance();
 	public static final String TRANSFER_SKIN = "PEApplySkinOnJoin";
 
+	//TODO: remove on player leave
+	private Map<Connection, String> connectionToSkinMap = new WeakHashMap<>();
```

Comment:
Why use connection map and not the uuid already contained inside the connection metadata?

Reply:
going to remove this cause it doesnt work anyways- did this before i merged in your last changes. but basically the uuid being stored was the skin uuid, and not the player uuid =\ so needed some association. the code you added looks better tho, so just gonna switch to that